### ISSUE TITLE
IntoEdges/IntoEdgesDirected for Reversed?

### ISSUE THREAD

**Issue by sunshowers** [petgraph/petgraph#292]
Hey --

I'm trying to write some generic edge-based graph algorithms, and it looks like I've run into a snag with the `Reversed` adapter for graphs. The adapter doesn't implement `IntoEdges` and `IntoEdgesDirected`, which is somewhat inconvenient :)

I tried giving an implementation a shot:

```
impl<G> IntoEdges for Reversed<G>
    where G: IntoEdgesDirected
{
    type Edges = ReversedEdges<G::EdgesDirected>;
    fn edges(self, a: Self::NodeId) -> Self::Edges {
        ReversedEdges {
            iter: self.0.edges_directed(a, Incoming),
        }
    }
}

impl<G> IntoEdgesDirected for Reversed<G>
    where G: IntoEdgesDirected
{
    type EdgesDirected = ReversedEdges<G::EdgesDirected>;
    fn edges_directed(self, a: Self::NodeId, dir: Direction) -> Self::Edges {
        ReversedEdges {
            iter: self.0.edges_directed(a, dir.opposite()),
        }
    }
}

/// A reversed edges iterator.
pub struct ReversedEdges<I> {
    iter: I,
}

impl<I> Iterator for ReversedEdges<I>
    where I: Iterator, I::Item: EdgeRef
{
    type Item = ReversedEdgeReference<I::Item>;
    fn next(&mut self) -> Option<Self::Item> {
        self.iter.next().map(ReversedEdgeReference)
    }
}
```

This appears to work for directed graphs but not for undirected ones. Here's why:

For a directed graph `g`:

```
0 --> 1 --> 2
```

consider `g.edges(1)`. `self.0.edges_directed(1, Incoming)` would return an edge with `source = 0, target = 1`. Reversing it with `ReversedEdgeReference` is the right thing to do.

For an *undirected* graph `g`:

```
0 -- 1 -- 2
```

consider `g.edges(1)`. `self.0.edges_directed(1, Incoming)` would return edges with `source = 1, target = 0`, as well as `source = 1, target = 2`. This means that reversing the edges with `ReversedEdgeReference` is the *wrong* thing to do.

Basically, this seems to be reflecting the idea that for an undirected graph, `Reversed` shouldn't do any transformations. But I don't see any way to do this in a generic fashion. Any ideas? :)

**Comment by sunshowers** [comment]
My current workaround is to copy-paste the implementation of `Reversed` into my own library and use the above impls for `IntoEdges` and `IntoEdgesDirected`, knowing up front that all my graphs are directed.

**Comment by sunshowers** [comment]
It seems to me that `self.edges_directed(node_id, Incoming)` for an *undirected* graph should return edges with target = node_id (i.e. use `swap_pair`), but that would probably be a breaking change.

**Comment by sunshowers** [comment]
Ah, looks like petgraph 0.5 is scheduled to be released soon (#246), and maybe this can be incorporated as part of that? I may put up a PR to change the behavior of `Incoming`.

**Comment by XVilka** [comment]
If you could send a PR with corresponding tests, that might become the part of 0.5, depending on how much it changes the code/behavior. Sorry for the inactivity, seems all maintainers are busy these months.

**Comment by sunshowers** [comment]
Thank you! I'm recovering from surgery right now, going to see if I have some time to do it soon.

**Comment by sunshowers** [comment]
Just upgraded our crate https://crates.io/crates/guppy to 0.5, works wonderfully. Thank you!
